src/Plugin/TfaBasePlugin.php

```diff
@@ -0,0 +1,187 @@
+<?php
+
+namespace Drupal\tfa\Plugin;
+
+use Drupal\Component\Plugin\PluginBase;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Base plugin class.
+ */
+abstract class TfaBasePlugin extends PluginBase {
+
+  /**
+   * @var string
+   */
+  protected $code;
+
+  /**
+   * @var int
+   */
+  protected $codeLength;
+
+  /**
+   * @var array
+   */
+  protected $context;
+
+  /**
+   * @var array
+   */
+  protected $errorMessages = array();
+
+  /**
+   * @var bool
+   */
+  protected $isValid;
+
+  /**
+   * @var string
+   */
+  protected $encryptionKey;
+
+  /**
+   * Constructs a TfaPlugin object.
+   *
+   * @param array $context
+   *   Context of current TFA process.
+   *   An array of paths keyed by their corresponding namespaces.
+   *
+   *   Must include key:
+   *
+   *     - 'uid'
+   *       Account uid of user in TFA process.
+   *
+   *   May include keys:
+   *
+   *     - 'validate_context'
+   *       Plugin-specific context for use during Tfa validation.
+   *
+   *     - 'setup_context'
+   *       Plugin-specific context for use during TfaSetup.
+   *
+   */
+  public function __construct(array $context) {
+    $this->context = $context;
+    // Default code length is 6.
+    $this->codeLength = 6;
+    $this->isValid = FALSE;
+  }
+
+  /**
+   * Determine if the plugin can run for the current TFA context.
+   *
+   * @return bool
+   */
+  public function ready() {
+    return TRUE;
+  }
+
+  /**
+   * Get error messages suitable for form_set_error().
+   *
+   * @return array
+   */
+  public function getErrorMessages() {
+    return $this->errorMessages;
+  }
+
+  /**
+   * Submit form.
+   *
+   * @param array $form
+   * @param FormStateInterface $form_state
+   * @return bool Whether plugin form handling is complete.
+   *   Plugins should return FALSE to invoke multi-step.
+   */
+  public function submitForm(array $form, FormStateInterface &$form_state) {
+    return $this->isValid;
+  }
+
+  /**
+   * Validate code.
+   *
+   * Note, plugins overriding validate() should be sure to set isValid property
+   * correctly or else also override submitForm().
+   *
+   * @param string $code Code to be validated
+   * @return bool Whether code is valid
+   */
+  protected function validate($code) {
+    if ((string) $code === (string) $this->code) {
+      $this->isValid = TRUE;
+      return TRUE;
+    }
+    else {
+      return FALSE;
+    }
+  }
+
+  /**
+   * Generate a random string of characters of length $this->codeLength.
+   *
+   * @return string
+   */
+  protected function generate() {
+    $characters = '123456789abcdefghijklmnpqrstuvwxyz';
+    $string = '';
+    $max = strlen($characters) - 1;
+    for ($p = 0; $p < $this->codeLength; $p++) {
+      $string .= $characters[mt_rand(0, $max)];
```

Comment:
Probably can just use something like this: http://php.net/manual/en/function.mt-rand.php